File: .github/helper/record_sorter.py
SECTIONS_TO_SORT = ["Apps", "Developer Tooling", "Deployment Tools", "Other Clients"]
NEWLINE = "\n"

def sort_section(lines: list[str]) -> list[str]:
    sorted_lines = []
    heading_buffer = []

    for line in lines:
        is_heading = line.startswith("#")
        is_list = line.startswith("-")

        if line == NEWLINE:
            continue

        if is_heading:
            heading_buffer.sort(key=str.casefold)
            sorted_lines.extend(heading_buffer + [NEWLINE])
            sorted_lines.append(line)
            heading_buffer = [NEWLINE]

        elif is_list:
            heading_buffer.append(line)

        else:
            if sorted_lines[-1] != NEWLINE:
                sorted_lines.append(NEWLINE)
            sorted_lines.append(line)

    if heading_buffer:
        heading_buffer.sort(key=str.casefold)
        sorted_lines.extend(heading_buffer + [NEWLINE])

    return sorted_lines


if __name__ == "__main__":
    README = open("README.md").readlines()
    README_sorted = README.copy()

    for section in SECTIONS_TO_SORT:
        section_start = README_sorted.index(f"### {section}\n")
        section_end = next(i for i in range(section_start + 1, len(README_sorted) + 1) if README_sorted[i].startswith("### "))
        README_sorted[section_start - 1:section_end] = sort_section(README_sorted[section_start:section_end])

    if README != README_sorted:
        open("README.md", "w").writelines(README_sorted)

File: .github/helper/record_sorter.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

SECTIONS_TO_SORT = ["Apps", "Developer Tooling", "Deployment Tools", "Other Clients"]
NEWLINE = "\n"

def sort_section(lines)
  sorted_lines = []
  heading_buffer = []

  lines.each do |line|
    is_heading = line.start_with?("#")
    is_list = line.start_with?("-")

    next if line == NEWLINE

    if is_heading
      heading_buffer.sort_by! { |l| l.downcase }
      sorted_lines.concat(heading_buffer + [NEWLINE])
      sorted_lines << line
      heading_buffer = [NEWLINE]
    elsif is_list
      heading_buffer << line
    else
      sorted_lines << NEWLINE unless sorted_lines.last == NEWLINE
      sorted_lines << line
    end
  end

  if !heading_buffer.empty?
    heading_buffer.sort_by! { |l| l.downcase }
    sorted_lines.concat(heading_buffer + [NEWLINE])
  end

  sorted_lines
end

if __FILE__ == $PROGRAM_NAME
  readme = File.readlines("README.md")
  readme_sorted = readme.dup

  SECTIONS_TO_SORT.each do |section|
    section_start = readme_sorted.index("### #{section}\n")
    next unless section_start

    section_end = (section_start + 1...readme_sorted.length).find { |i| readme_sorted[i].start_with?("### ") }
    section_end ||= readme_sorted.length

    readme_sorted[section_start - 1...section_end] = sort_section(readme_sorted[section_start...section_end])
  end

  if readme != readme_sorted
    File.write("README.md", readme_sorted.join)
  end
end
